library/rmc/src/invariant.rs

```diff
@@ -0,0 +1,162 @@
+        byte == 0 || byte == 1
+    }
+}
+
+unsafe impl Invariant for u8 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for u16 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for u32 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for u64 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for u128 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for usize {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for i8 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for i16 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for i32 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for i64 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for i128 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+/// We do not constraint floating points values per type spec. Users must add assumptions to their
+/// verification code if they want to eliminate NaN, infinite, or subnormal.
+unsafe impl Invariant for f32 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+/// We do not constraint floating points values per type spec. Users must add assumptions to their
+/// verification code if they want to eliminate NaN, infinite, or subnormal.
+unsafe impl Invariant for f64 {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+unsafe impl Invariant for isize {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+/// Verifies that:
+/// - a char is not outside the ranges [0x0, 0xD7FF] and [0xE000, 0x10FFFF]
+/// Ref: https://doc.rust-lang.org/stable/nomicon/what-unsafe-does.html
+unsafe impl Invariant for char {
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        // RMC translates char into i32.
+        let val = *self as i32;
+        val <= 0xD7FF || (val >= 0xE000 && val <= 0x10FFFF)
+    }
+}
+
+unsafe impl<T> Invariant for Option<T>
+where
+    T: Invariant,
+{
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        if let Some(v) = self { v.is_valid() } else { matches!(*self, None) }
+    }
+}
+
+unsafe impl<T, E> Invariant for Result<T, E>
+where
+    T: Invariant,
+    E: Invariant,
+{
+    #[inline(always)]
+    fn is_valid(&self) -> bool {
+        if let Ok(v) = self {
+            v.is_valid()
+        } else if let Err(e) = self {
+            e.is_valid()
+        } else {
+            false
+        }
+    }
+}
```

Comment:
The `assumptions` module handled a lot more types than the ones implemented here, e.g. enums. Do we lose assumptions for those other types?

Reply:
Yes, we do. For enums and structs, @danielsn and I were thinking about adding a custom derive: #726

For other cases, I would prefer being safe. I don't think we should infer anything based on the type of fields without any sort of user intervention.

Reply:
My two cents: it would be inconvenient if the user needs to annotate all the enums in their code with `#[derive(Invariant)]`. Automatically handling that during codegen is more user friendly.